das/worker.go

```diff
@@ -0,0 +1,93 @@
+package das
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+)
+
+type worker struct {
+	lock     sync.Mutex
+	state    workerState
+	resultCh chan<- result
+	sample   func(context.Context, uint64) error
+}
+
+// workerState contains important information about the state of a
+// current sampling routine.
+type workerState struct {
+	Curr uint64 `json:"curr"`
+	From uint64 `json:"from"`
+	To   uint64 `json:"to"`
+
+	failed []uint64
+	Err    error `json:"error,omitempty"`
+}
+
+type job struct {
+	id         int
+	from, to   uint64
+	isPriority bool
+}
+
+func newWorker(
+	resultCh chan<- result,
+	sample func(context.Context, uint64) error) worker {
+	return worker{
+		resultCh: resultCh,
+		sample:   sample,
+	}
+}
+
+func (w *worker) run(ctx context.Context, j job) {
+	w.setStateFromJob(j)
+
+	for curr := j.from; curr <= j.to; curr++ {
+		err := w.sample(ctx, curr)
+		w.setResult(curr, err)
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
```

Comment:
Main reason is to have `curr` iterator at latest attempt before context got canceled.  Further it is stored stored in checkpoint and after restart worker will resume from same `curr`. Otherwise all headers that are left in job after `cancel()` call will be stored as error.

`Done()` brings very little overhead because it is `<-chan struct{}` and should not affect other parallel `Done()` calls

Reply:
> Done() brings very little overhead because it is <-chan struct{} and should not affect other parallel Done() calls

I meant the mutex inside the Done call

> Main reason is to have curr iterator at latest attempt before context got canceled. Further it is stored stored in checkpoint and after restart worker will resume from same curr. Otherwise all headers that are left in job after cancel() call will be stored as error.

Ok, so there is a correctness issue here.

Reply:
This makes me think about retrying for failed headers

Reply:
Which is actually here. This all takes time to absorb and chew

Reply:
Yeah, it is sort of hidden inside save/load checkpoint functionality.